### PR TITLE
WT-11280 Upgrading the barrier in __wt_session_gen_enter

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by dgomezferro — CHANGES_REQUESTED** [review]
I believe [we need to make this oldest = conn->generations[which] read ](https://github.com/wiredtiger/wiredtiger/blob/e3cfaefe392309b44a9f7d79bfff317361cc0a9b/src/support/generation.c#L235) a WT_ORDERED_READ too.

File: tools/litmus_tests/generations/gen_drain_arm64.litmus
Comment (luke-pearson) [inline tools/litmus_tests/generations/gen_drain_arm64.litmus:1]: @dgomezferro I tried to come up with this one but haven't yet, would you be able to assist? This is what I got:
```
AArch64 wt_gen_drain
{
conn_gen=5;
0:X1=conn_gen; 0:X3=sess_gen;
1:X1=conn_gen; 1:X3=sess_gen;
}

 P0                   | P1             ;
 MOV W9, #1           | LDR W2, [X1]   ;
 ldaddal W9, W1, [X1] | STR W2, [X3]   ;
 LDR W0, [X3]         | dmb ish        ;
                      | LDR W0, [X3]   ;

exists (0:X3=0 /\ 1:X3=5 /\ 1:X1=5)
```

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/support/generation.c [hunk 1]
```diff
@@ -232,7 +232,15 @@ __gen_oldest(WT_SESSION_IMPL *session, int which)
      * the sessions that could have been active when we started our check.
      */
     WT_ORDERED_READ(session_cnt, conn->session_cnt);
-    for (oldest = conn->generations[which], s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
+    /*
+     * On platforms with weaker memory models we need to order the read of the connection generation
+     * before the read of the session generation. If the session generation read is ordered before
```
Comment (dgomezferro) [inline src/support/generation.c:237]: I would omit the mention of weaker memory models, we need this specific ordering, some platforms guarantee it and some don't, I think that's WT_ORDERED_READ's job, to do something or nothing depending on which platform we are building for.

File: tools/litmus_tests/generations/gen_oldest_arm64.litmus [hunk 2]
```diff
@@ -0,0 +1,15 @@
+AArch64 wt_gen_oldest
+{
+conn_gen=5;sess_gen=0;
+0:X1=conn_gen; 0:X3=sess_gen;
+1:X1=conn_gen; 1:X3=sess_gen;
+2:X1=conn_gen; 2:X3=sess_gen;
+}
+
+ P0            | P1           | P2                   ;
+ LDR W0, [X1]  | LDR W0, [X1] | MOV W9, #1           ;
```
Comment (dgomezferro) [inline tools/litmus_tests/generations/gen_oldest_arm64.litmus:10]: We could simplify this using X registers everywhere and less random register numbers if you wish.

File: tools/litmus_tests/README.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# WiredTiger litmus tests
```
Comment (dgomezferro) [inline tools/litmus_tests/README.md:1]: Nice idea adding them to the codebase!

File: tools/litmus_tests/generations/gen_drain_x86.litmus [hunk 4]
```diff
@@ -0,0 +1,12 @@
+X86 wt_gen_drain
+{
+conn_gen=5;
+}
+
+ P0                       | P1                    ;
+ LOCK; ADD [conn_gen], $1 | MOV EAX,[conn_gen]    ;
+ LFENCE                   | MOV [sess_gen], EAX   ;
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/litmus_tests/generations/gen_drain_x86.litmus:8]: Yeah this LFENCE is the `conn->session_cnt` LFENCE. I'll remove it.

File: tools/litmus_tests/generations/gen_drain_arm64.litmus
Comment (luke-pearson) [reply tools/litmus_tests/generations/gen_drain_arm64.litmus:1]: Thanks Dani, I'll push up the clearer version.

File: tools/litmus_tests/generations/gen_oldest_arm64.litmus (hunk 2 above)
Comment (luke-pearson) [reply tools/litmus_tests/generations/gen_oldest_arm64.litmus:10]: I'll push up a simplification

File: tools/litmus_tests/README.md (hunk 3 above)
Comment (luke-pearson) [reply tools/litmus_tests/README.md:1]: Thanks!

File: tools/litmus_tests/generations/gen_drain_x86.litmus (hunk 4 above)
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/litmus_tests/generations/gen_drain_x86.litmus]: [ea517a1](https://github.com/wiredtiger/wiredtiger/pull/9463/commits/ea517a17db053aa41a298195b7daec62607afe4e)

File: src/support/generation.c (hunk 1 above)
Comment (luke-pearson) [reply src/support/generation.c]: [ea517a1](https://github.com/wiredtiger/wiredtiger/pull/9463/commits/ea517a17db053aa41a298195b7daec62607afe4e)

File: tools/litmus_tests/generations/gen_drain_arm64.litmus
Comment (luke-pearson) [reply tools/litmus_tests/generations/gen_drain_arm64.litmus:1]: [ea517a1](https://github.com/wiredtiger/wiredtiger/pull/9463/commits/ea517a17db053aa41a298195b7daec62607afe4e)

File: tools/litmus_tests/generations/gen_oldest_arm64.litmus (hunk 2 above)
Comment (luke-pearson) [reply tools/litmus_tests/generations/gen_oldest_arm64.litmus]: [5b86b30](https://github.com/wiredtiger/wiredtiger/pull/9463/commits/5b86b3009c1ebe9d6b8bcc22a5a477de4c1645e1)

**Review by agorrod — APPROVED** [review]
Thanks @luke-pearson - the litmus test is a great addition. Very cool.

**Review by dgomezferro — APPROVED** [review]
LGTM, nice comments!

**Pull request merged**